src/compiler.jl

```diff
@@ -714,66 +1054,287 @@ end
+        end
+    end
+    wrapped = Expr[]
+    for i in 1:N
+        expr = :(
+                if ActivityTup[$i+1] && !isghostty($(primtypes[i]))
+                if !$forwardMode && ($(primtypes[i]) <: AbstractFloat || $(primtypes[i]) <: Complex{<:AbstractFloat})
+                    Active($(primargs[i]))
+                 else
+                     $((Width == 1) ? :Duplicated : :BatchDuplicated)($(primargs[i]), $(shadowargs[i]))
+                 end
+             else
+                 Const($(primargs[i]))
+             end
+
+            )
+        push!(wrapped, expr)
+    end
+    return primargs, shadowargs, primtypes, allargs, typeargs, wrapped
+end
+
+const FwdCache = Dict{Tuple{Int64,Int64},Function}()
+function runtime_generic_fwd(N::Int64,Width::Int64)
+    if haskey(FwdCache, (N,Width))
+        return FwdCache[(N,Width)]
+    end
+    primargs, shadowargs, primtypes, allargs, typeargs, wrapped = setup_macro_wraps(true, N, Width)
+    funcname = Symbol("inner_runtime_generic_fwd_"*string(N)*"_"*string(Width))
+    efn = quote
+        Base.@constprop :aggressive function $funcname(activity::Val{ActivityTup}, width::Val{$Width}, RT::Val{ReturnType}, f::F, df::DF,$(allargs...)) where {ActivityTup,ReturnType, F, DF, $(typeargs...)}
+            args = ($(wrapped...),)
+
+            fn = f
+            dfn = ActivityTup[1] ? df : nothing
+
+            # TODO: Annotation of return value
+            # tt0 = Tuple{$(primtypes...)}
+            tt = Tuple{map(x->eltype(typeof(x)), args)...}
+            tt′ = Tuple{map(typeof, args)...}
+            rt = Core.Compiler.return_type(fn, tt)
+            annotation = guess_activity(rt, API.DEM_ForwardMode)
+            
+            if annotation <: DuplicatedNoNeed
+                annotation = Duplicated{rt}
+            end
+            if $Width != 1
+                if annotation <: Duplicated
+                    annotation = BatchDuplicated{rt, Width}
+                end
+            end
+
+            forward = thunk(fn, dfn, annotation, tt′, Val(API.DEM_ForwardMode), width, #=ModifiedBetween=#Val(false), #=returnPrimal=#Val(true))
+
+            res = forward(args...)
+
+            if length(res) == 0
+                return ReturnType(ntuple(i->nothing, Val($Width+1)))
+            end
+            if annotation <: Const
+                return ReturnType(ntuple(i->res[1], Val($Width+1)))
+            end
+
+            if $Width == 1
+               return ReturnType((res[1], res[2]))
             else
-                d[w]
+                return ReturnType((res[1], res[2]...))
             end
-            a = unsafe_load(shadow_ptr, (i-1)*Width+w)
-            if a === nothing
-            elseif a isa Base.RefValue
-                @assert eltype(a) == typeof(dd)
-                a[] += dd
+        end
+    end
+    fn = eval(efn)
+    FwdCache[(N,Width)] = fn
+    return fn
+
+end
+
+const AugCache = Dict{Tuple{Int64,Int64},Function}()
```

Comment:
Let's add tests that show that this is correct under invalidation of functions. Not that we need to bifurcate this cache on a world-age range.

Reply:
So this cache is not type dependent, but arg/width, dependent. It contains the generic templated function for a given arg/width combo.

The bigger concern is that if you look at the impl it requires an eval (which ideally would be generated on the fly, but per an eval in a generated function being illegal, I just run a for loop over "common" sizes). We should also add a fallback to the current fastgen2 style vararg code if its outside that, unless we have a way to get around the eval in generated.

Reply:
Can we use an OpaqueClosure instead? Or just an anonymous function?

Reply:
sure, re either